src/rapids_singlecell/_cuda/aggr/aggr.cu

```diff
@@ -0,0 +1,171 @@
+                             (cudaStream_t)stream);
+    }
+  } else {
+    if (dtype_itemsize == 4) {
+      launch_dense_C<float>(data, out, cats, mask, n_cells, n_genes, n_groups,
+                            (cudaStream_t)stream);
+    } else {
+      launch_dense_C<double>(data, out, cats, mask, n_cells, n_genes, n_groups,
+                             (cudaStream_t)stream);
+    }
+  }
+}
+
+static inline void csr_to_coo_dispatch(std::uintptr_t indptr, std::uintptr_t index,
+                                       std::uintptr_t data, std::uintptr_t row, std::uintptr_t col,
+                                       std::uintptr_t ndata, std::uintptr_t cats,
+                                       std::uintptr_t mask, int n_cells, int dtype_itemsize,
+                                       std::uintptr_t stream) {
+  if (dtype_itemsize == 4) {
+    launch_csr_to_coo<float>(indptr, index, data, row, col, ndata, cats, mask, n_cells,
+                             (cudaStream_t)stream);
+  } else {
+    launch_csr_to_coo<double>(indptr, index, data, row, col, ndata, cats, mask, n_cells,
+                              (cudaStream_t)stream);
+  }
+}
+
+// variance launcher
+static inline void launch_sparse_var(std::uintptr_t indptr, std::uintptr_t index,
+                                     std::uintptr_t data, std::uintptr_t mean_data,
+                                     std::uintptr_t n_cells, int dof, int n_groups,
+                                     cudaStream_t stream) {
+  dim3 grid((unsigned)n_groups);
+  dim3 block(64);
+  sparse_var_kernel<<<grid, block, 0, stream>>>(
+      reinterpret_cast<const int*>(indptr), reinterpret_cast<const int*>(index),
+      reinterpret_cast<double*>(data), reinterpret_cast<const double*>(mean_data),
+      reinterpret_cast<double*>(n_cells), dof, n_groups);
+}
+
+NB_MODULE(_aggr_cuda, m) {
+  m.def("sparse_aggr", &sparse_aggr_dispatch, "indptr"_a, "index"_a, "data"_a, nb::kw_only(),
+        "out"_a, "cats"_a, "mask"_a, "n_cells"_a, "n_genes"_a, "n_groups"_a, "is_csc"_a,
+        "dtype_itemsize"_a, "stream"_a = 0);
+  m.def("dense_aggr", &dense_aggr_dispatch, "data"_a, nb::kw_only(), "out"_a, "cats"_a, "mask"_a,
+        "n_cells"_a, "n_genes"_a, "n_groups"_a, "is_fortran"_a, "dtype_itemsize"_a, "stream"_a = 0);
+  m.def("csr_to_coo", &csr_to_coo_dispatch, "indptr"_a, "index"_a, "data"_a, nb::kw_only(),
+        "out_row"_a, "out_col"_a, "out_data"_a, "cats"_a, "mask"_a, "n_cells"_a, "dtype_itemsize"_a,
+        "stream"_a = 0);
+  m.def(
+      "sparse_var",
+      [](std::uintptr_t indptr, std::uintptr_t index, std::uintptr_t data, std::uintptr_t mean_data,
+         std::uintptr_t n_cells, int dof, int n_groups, std::uintptr_t stream) {
+        launch_sparse_var(indptr, index, data, mean_data, n_cells, dof, n_groups,
+                          (cudaStream_t)stream);
+      },
+      "indptr"_a, "index"_a, "data"_a, nb::kw_only(), "means"_a, "n_cells"_a, "dof"_a, "n_groups"_a,
+      "stream"_a = 0);
```

Comment:
Please add [`nb::kw_only`](https://nanobind.readthedocs.io/en/latest/api_core.html#_CPPv4N8nanobind7kw_onlyE) like I did here, so only the inputs are positional

- if the input is an array, that’s 1 positional argument
- if the input is a matrix, that’s 3 positional arguments
- if the function is a binary operation on 2 arrays, that’s 2 positional arguments
- if the function is a binary operation on 2 matrices, that’s 6 positional arguments

out arguments should always be called `out` (or `out_indptr`, `out_indices` and `out_data` for matrices).